### PR TITLE
Persistence Component: Introduce Fail on Duplicate Primary Keys

### PR DESCRIPTION
**Pull request by kumuwu**
#### What type of PR is this?

- Improvement

#### What does this PR do / why is it needed ?

Introduced new boolean: failOnDuplicatePrimaryKeys under NoVersioning

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/unitemporal/UnitmemporalDeltaBatchIdDateTimeBasedTestCases.java
Comment (rengam32) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/unitemporal/UnitmemporalDeltaBatchIdDateTimeBasedTestCases.java:1]: Unrelated to your change - but can we rename the file to correct the typo

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ErrorCategory.java [hunk 1]
```diff
@@ -23,6 +23,7 @@ public enum ErrorCategory
     FILE_NOT_FOUND("File not found in specified location"),
     UNKNOWN("Unknown error"),
     DUPLICATES("Duplicate rows found"),
+    DUPLICATE_PRIMARY_KEYS("Duplicate primary keys found"),
```
Comment (rengam32) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ErrorCategory.java:26]: Make the error mss bit more intuitive 
"Multiple rows with duplicate primary keys found"

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java [hunk 2]
```diff
@@ -470,6 +472,24 @@ public void dedupAndVersion()
...
+            if (dedupAndVersionErrorSqlTypeSqlPlanMap.containsKey(MAX_PK_DUPLICATES))
+            {
+                List<TabularData> result = executor.executePhysicalPlanAndGetResults(dedupAndVersionErrorSqlTypeSqlPlanMap.get(MAX_PK_DUPLICATES));
+                Optional<Object> obj = getFirstColumnValue(getFirstRowForFirstResult(result));
+                Optional<Long> maxPkDuplicatesValue = retrieveValueAsLong(obj.orElse(null));
+                if (maxPkDuplicatesValue.isPresent() && maxPkDuplicatesValue.get() > 1)
+                {
+                    // Find the pk-duplicate rows
+                    TabularData duplicatePkRows = executor.executePhysicalPlanAndGetResults(dedupAndVersionErrorSqlTypeSqlPlanMap.get(PK_DUPLICATE_ROWS)).get(0);
+                    String errorMessage = "Encountered duplicate primary keys, Failing the batch as Fail on Duplicate Primary Keys is selected";
```
Comment (rengam32) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java:485]:
```suggestion
                    String errorMessage = "Encountered multiple rows with duplicate primary keys, Failing the batch as Fail on Duplicate Primary Keys is selected";
```

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java [hunk 3]
```diff
@@ -470,6 +472,24 @@ public void dedupAndVersion()
...
+                    LOGGER.error(errorMessage);
```
Comment (rengam32) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java:486]: Maybe the logger also contain some details about the running pipeline? (which dataset/datastore/ etc) will be useful when debugging

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/unitemporal/UnitmemporalDeltaBatchIdDateTimeBasedTestCases.java
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-test/src/test/java/org/finos/legend/engine/persistence/components/testcases/ingestmode/unitemporal/UnitmemporalDeltaBatchIdDateTimeBasedTestCases.java:1]: Sure - will keep this in a note and do it in a later PR. As we aim to merge a PRs in one go this time, maybe best to avoid renaming of files which may have rebase issues. But yes will do it soon

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java (hunk 2 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java:485]: Sure, changed!

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java (hunk 3 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java:486]: It seems like the library loggers traditionally do not log these info. LH already logs it so we already have this information for debugging

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/versioning/TestDedupAndVersioning.java
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/versioning/TestDedupAndVersioning.java:1]: This is great, Can we also add one test in H2 Ingest test as well preferable for UnitempDelta
like we have for other dedup combinations ? 
To save time, We can also consider to update one of the existing tests and last pass could be duplicate PKs.

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/ingestmode/versioning/DeriveDuplicatePkRowsLogicalPlan.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                .addValue(All.INSTANCE)
+                .alias(DUPLICATE_PK_COUNT)
+                .build();
+
+            Selection selectDuplicatePks = Selection.builder()
+                .source(tempStagingDataset)
+                .groupByFields(pks)
+                .addAllFields(pks)
+                .addFields(count)
+                .havingCondition(GreaterThan.of(FieldValue.builder().fieldName(DUPLICATE_PK_COUNT).build(), ObjectValue.of(1)))
```
[the remark this answers is not in the document]
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/ingestmode/versioning/DeriveDuplicatePkRowsLogicalPlan.java:64]: Yes we cannot use "having" in postgres. However that part of the change has not been merged - we need to introduce a new Capability to achieve that (see: https://github.com/finos/legend-engine/pull/2775). Since postgres is not prioritized/used for now, can we do the refactoring in the postgres PR instead?

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/versioning/TestDedupAndVersioning.java
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/versioning/TestDedupAndVersioning.java:1]: Sure, added in unitemporal delta a batch that would fail the ingestion process because of duplicate PKs. (The happy path where we enable the flag but no duplicate PKs exist have been tested across different ingest modes)

Comment (prasar-ashutosh) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-h2/src/test/java/org/finos/legend/engine/persistence/components/versioning/TestDedupAndVersioning.java:1]: Great, thanks

**Review by epsstan — APPROVED** [review]
